### ISSUE TITLE
make insert() executemany() support RETURNING for as many dialects as possible, usually with VALUES() 

### ISSUE THREAD

**Issue by zzzeek** [sqlalchemy/sqlalchemy#6047]
see https://github.com/sqlalchemy/sqlalchemy/issues/6047#issuecomment-1081968916 for current thinking on this.

old:

looking at the code at:

https://github.com/psycopg/psycopg2/blob/master/lib/extras.py#L1212

and:

https://github.com/sqlalchemy/sqlalchemy/blob/master/lib/sqlalchemy/dialects/postgresql/psycopg2.py#L855

we create a generic version of execute_batch() / execute_values(), probably one that works with the string as unicode, then move the psycopg2 logic down to postgresql/base.py.

The asyncpg driver is likely going to be very popular so based on comments like https://github.com/sqlalchemy/sqlalchemy/discussions/6045#discussioncomment-480654 this is likely going to keep coming up.

**Comment by zzzeek** [comment]
also:

* sqlite if we can add returning as in #6195
* does SQL server support VALUES () () () ?  
* Oracle in fact can support multi-row RETURNING also, so we'd want to implement that and then see if Oracle can get it as well
* MariaDB supports both syntaxes

It would be nice if this could move into DefaultExecutionContext / Connection in some way where we can actually log the SQL and parameters being run as they actually are.   maybe like this:

```diff
diff --git a/lib/sqlalchemy/engine/base.py b/lib/sqlalchemy/engine/base.py
index 0a3ce3bdc..0c24527f7 100644
--- a/lib/sqlalchemy/engine/base.py
+++ b/lib/sqlalchemy/engine/base.py
@@ -1580,6 +1580,28 @@ class Connection(Connectable):
             future=True,
         )
 
+    def _log_statement(self, context, statement, parameters):
+        self._log_info(statement)
+
+        stats = context._get_cache_stats()
+
+        if not self.engine.hide_parameters:
+            self._log_info(
+                "[%s] %r",
+                stats,
+                sql_util._repr_params(
+                    parameters,
+                    batches=10,
+                    ismulti=context.executemany
+                    and not context.fastexecutemany,
+                ),
+            )
+        else:
+            self._log_info(
+                "[%s] [SQL parameters hidden due to hide_parameters=True]"
+                % (stats,)
+            )
+
     def _execute_context(
         self,
         dialect,
@@ -1653,26 +1675,6 @@ class Connection(Connectable):
                     context.executemany,
                 )
 
-        if self._echo:
-
-            self._log_info(statement)
-
-            stats = context._get_cache_stats()
-
-            if not self.engine.hide_parameters:
-                self._log_info(
-                    "[%s] %r",
-                    stats,
-                    sql_util._repr_params(
-                        parameters, batches=10, ismulti=context.executemany
-                    ),
-                )
-            else:
-                self._log_info(
-                    "[%s] [SQL parameters hidden due to hide_parameters=True]"
-                    % (stats,)
-                )
-
         evt_handled = False
         try:
             if context.executemany:
@@ -1682,9 +1684,19 @@ class Connection(Connectable):
                             evt_handled = True
                             break
                 if not evt_handled:
-                    self.dialect.do_executemany(
-                        cursor, statement, parameters, context
-                    )
+                    if context.fastexecutemany:
+                        for stmt, batch in context.fast_exec_batches:
+                            if self._echo:
+                                self._log_statement(context, stmt, batch)
+                            self.dialect.do_execute(
+                                cursor, stmt, batch, context
+                            )
+                    else:
+                        if self._echo:
+                            self._log_statement(context, statement, parameters)
+                        self.dialect.do_executemany(
+                            cursor, statement, parameters, context
+                        )
             elif not parameters and context.no_parameters:
                 if self.dialect._has_events:
                     for fn in self.dialect.dispatch.do_execute_no_params:
@@ -1692,6 +1704,8 @@ class Connection(Connectable):
                             evt_handled = True
                             break
                 if not evt_handled:
+                    if self._echo:
+                        self._log_statement(context, statement, parameters)
                     self.dialect.do_execute_no_params(
                         cursor, statement, context
                     )
@@ -1702,6 +1716,8 @@ class Connection(Connectable):
                             evt_handled = True
                             break
                 if not evt_handled:
+                    if self._echo:
+                        self._log_statement(context, statement, parameters)
                     self.dialect.do_execute(
                         cursor, statement, parameters, context
                     )

```

this moves the logging to be per-statement and after event handlers.   this would also allow an event-handler-modified statement to be logged in its final form.    it would remove the behavior of the statement being logged for an event handler that cancels the execution of a statement, which is also likely more appropriate.

**Comment by zzzeek** [comment]
oracle multirow returning in #6245 

my current thinking here is:

1. don't move around the logging and all that for now, too much change mid-release
2. build a new dialect method do_execute_insert_batch().  maybe add some additional logging around this method but that's it.
3. all supporting dialects enable it using "executemany_mode="values_only"" to the base.
4. turn on by default for postgresql dialects, off by default for others.
5. suite test will need to use a testing_engine that turns it on so we can test

**Comment by gordthompson** [comment]
> does SQL server support VALUES () () ()

Yes. They call it a "table value constructor" (TVC) and last time I checked it could insert up to 1_000 rows at a time.

**Comment by gordthompson** [comment]
> does SQL server support VALUES () () ()

Yes, they call it a "table value constructor" (TVC). However, they are limited in that:

1. A TVC can only insert a maximum of 1_000 rows under any circumstances, and
2. A parameterized TVC uses a SQL Server stored procedure and T-SQL SPs are limited to 2_100 parameters, so the number of columns can further reduce the TVC's 1_000-row limit. For example, for three (3) columns a parameterized TVC cannot insert more than 699 rows.

`fast_executemany = True` is only limited by available memory and is significantly faster than a TVC. Testing with raw pyodbc on a LAN-connected SQL Server instance:

```python
table_name = "##tmp"
create_table = f"CREATE TABLE {table_name} (id int, txt varchar(50), x float)"
crsr.execute(create_table)

num_rows = 699
# list of tuples:
data = [(i, f"record{i:03d}", 3.14159 * i) for i in range(num_rows)]
num_cols = len(data[0])


def use_tvc():
    placeholders = ",".join(
        [f"({','.join('?' * num_cols)})" for x in range(num_rows)]
    )
    sql = f"INSERT INTO {table_name} (id, txt, x) VALUES {placeholders}"

    # flatten to plain tuple:
    data_tuple = tuple(item for row in data for item in row)
    t0 = time.perf_counter()
    crsr.execute(sql, data_tuple)
    cnxn.commit()
    print(
        f"TVC: {num_rows} rows inserted in {(time.perf_counter() - t0):0.2f} seconds"
    )


def use_fast_executemany():
    crsr.fast_executemany = True
    t0 = time.perf_counter()
    crsr.executemany(
        f"INSERT INTO {table_name} (id, txt, x) values (?,?,?)", data
    )
    cnxn.commit()
    print(
        f"fast_executemany: {num_rows} rows inserted in {(time.perf_counter() - t0):0.2f} seconds"
    )


use_fast_executemany()
# fast_executemany: 699 rows inserted in 0.13 seconds

crsr.execute(f"DROP TABLE {table_name}")
crsr.execute(create_table)

use_tvc()
# TVC: 699 rows inserted in 0.52 seconds
```

**Comment by zzzeek** [comment]
OK but by running the staements ourselves with RETURNING we get the PK values back.   we cant do that with normal executemany, so for he ORM this would still be a huge speed improvement when we want the PKs back.  also the fast_executemany appears to suck up all the RAM without any way to control it, seemed buggy.

**Comment by zzzeek** [comment]
.13 sec vs. .52 is not as big a difference as .52 sec vs. 8 seconds if we use separate INSERT statements to get the PKs back

**Comment by zzzeek** [comment]
anyway, if they have fast_executemany set on their driver then this feature would not kick in for SQL server

**Comment by zzzeek** [comment]
note #6581 as well, this needs to happen only for a plain vanilla INSERT and not with any ON CONFLICT or anything special, we should make sure to lock it down.

**Comment by zzzeek** [comment]
additionally, executemanymode should be available as an execution option.

**Comment by zzzeek** [comment]
continue to agree this should not impact logging or the before_cursor_execute() event for now, those will still see normal "executemany" and the rewriting here will take place as a dialect-level optimization.

for dialect we should probably have a new method as mentioned with a new event, like do_execute_batches.

I think the name "executemany_mode" might be retired here in favor of something more specific to INSERTs, like enable_insert_batching=True or something like that (it would need to start as "auto" or something so that dialects can choose if it's not present, but then it can be turned on or off).   it also might be nice as an execution option.

I think the other psycopg2 "fast executemany" modes should just be removed for now.  I dont think anyone is using these explicitly.  the flag can emit a warning if it's set that it no longer does anything.   IIRC there were issues with executemany="batch" that made it not very useful but it may be good to revisit what those were.

also for asyncpg/ psycopg3 we need to make sure to explciitly tell cursor.execute() to not cache a prepared statement as these will be INSERT statements that have a huge number of parameters that varies each time.  we may want to say prepared caching can stay on for all batches that arent the last one, since these will have a fixed number of slots and be repeatable.

**Comment by zzzeek** [comment]
update, for non-psycopg2 drivers we should look to enabling this only at most for the ORM multi-returning use case.   asyncpg for example is now benched as doing straight "executemany()" without returning as much faster than chunked values().

**Comment by CaselIT** [comment]
I believe we should look into psycopg3 too, since iirc the issue with executemany was mainly a psycopg2 thing

**Comment by zzzeek** [comment]
okey doke I am experiencing a shift in thinking on this.

the insert() construct itself is something it looks like we will be [generalizing](https://github.com/sqlalchemy/sqlalchemy/issues/7865) as an ORM construct.   In this usage we expect users to be making a lot of use of insert().values(), with or without returning().

in addition, as noted above, "psycopg2 fast execution" seems like, at least outside of the RETURNING use case, something that is pretty much a psycopg2 issue.  that is, asyncpg runs better using normal executemany() if you aren't trying to get RETURNING.   lets assume psycopg3 does as well.  This changes the whole calculus of this quite a lot, in that there is much less need to generalize psycopg2's helpers; we dont need them unless we want to optimize the use of returning().  But when returning() is used, we can just do insert().values() anyway.

third, the ORM has explicit logic for when it does "executemany with returning", it calls upon this new accessor inserted_primary_key_rows in a total special case where it expects "executemany" to implicitly turn the statement into a series of chunked insert().values().   it is already complicated inside the ORM to use the important part of this feature.

it seems things would be clearer and more extensible if we had the ORM instead simply use insert().values().returning() directly, within the scope of this "special case" logic.  that way it's making use of a clear API that we are also providing to end users already.  The way 1.4 is stuffing this logic inside of "do_executemany()" is much more opaque and difficult to explain.

Another issue is that insert().values() is not currently cacheable, nor is the Values() construct, because these constructs represent a large number of variable entries that would not be appropriate for caching.    But in our fast execution helper system, the dialect is taking a vanilla insert() and doing string transformations on the compiled SQL.      This situation indicates a clear use case for the existing "expanding SQL" feature, which we use currently for IN expressions as well as some forms of "literals".    if insert.values() and Values() made use of a new "expanding SQL" feature, these constructs would then become cacheable, there would be an existing well-travelled path by which they get their parameters "expanded" at execute time, we could remove/deprecate weird accessors like inserted_primary_key_rows and keep the whole thing explicit. 

the remaining aspect of it is that when one uses insert().values().returning() for bulk insert, you probably want to batch it in sets of 1000 records or something like that.   the batching is extremely simple but is still an extra step.  the ORM can do this batching internally using a simple helper function of some kind.  But at the level of session.execute(), we might want to make ORM-enabled inserts (as we are adding in #7865 in any case) do this batching automatically.  There is already an ORM contextual object that can do this in place.

so to sum up:

- [ ] leave psycopg2 fast execution helpers as they are
- [ ] add a new "expanding" feature to insert().values(), Values(), so that these are cacheable and the DefaultExecutionContext can "expand" out the current statement's parameters on the fly the same way it already does for expanding bound parameters and other "post compile" parameters.   that is also all logic that can benefit by going to cython someday.   this allows all insert() and Values() constructs to be fully cachable.
- [ ] have the ORM use insert().values().returning() explicitly within unit of work, for all backends that can support the syntax.  that way there is no no change to the dialects needed here, no change to do_executemany() or any kind of weird tinkering inside of the Connection as I had previously suggested above, no need for weird new event handlers.
- [ ] as we fill out ORM-enabled insert(), see if it can do batching automatically.

**Comment by CaselIT** [comment]
Nice update. I think we could revisit the usefulness of vendoring that logic of psycopg with more calm for a 2.x release

**Comment by zzzeek** [comment]
noted at #7907, defaultdialect should emit a warning when the compiled statement has `_returning` present but we are doing an executemany().   it would need to take into account the fact that psycopg2 dialect **can** return rows and not warn in that case.  awkward

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **main** branch:

**wip: notes on using insert.values(....) for ORM directly** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3890

**Comment by zzzeek** [comment]
for reference, 92cc232726a01dd3beff762ebccd326a9659e8b9 implemented for #3288 adds default support for insert().values(), and I'm actually surprised i did it that way, it's doing the worst thing possible which is running the defaults over and over again all through the "single" set of parameters.

what i think has to happen is defaultexecutioncontext still needs a third, distinct style of execution, which is an "executemany single statement" kind of thing.  We should have compiled_params be a list, and handle defaults etc. having params as a list as though we're doing executemany, then at invocation time we do string / parameter manipulation to make it a single execute.  basically exactly what we are doing for psycopg2 fast execution, but at the default dialect level.

So.....the next thing that this brings us towards, is that there's not really much difference between `execute(insert.values( multiparams )) and `execute(insert, multiparams)` except for this arbitrary decision of how we want to organize the INSERT statement.

So.     I think we should also do this:

1. add an execution option such that `conn.execution_options(collapse_inserts=True).execute(insert, multiparams)` will do the same thing as `conn.execute(insert.values(multiparams))`
2. default behaviors remain that insert.values() always does a collapsed, single statement, `conn.execute(insert, multiparams)` uses executemany.    psycopg2 is likely the only driver that has a performance issue w/ the latter, if RETURNING is not used.  if RETURNING is used, then you need to do the "collapsed" form if available.
3. "collapsed" execution will populate context.compiled_params with a parameter set per row, even if we will be using execute(insert.values())
4. all default-running logic works in executemany() style
5. integrate multi-row RETURNING into inserted_primary_key_rows, returned_default_rows.  This already works except the former needs the list form of compiled_params to be there, so this should just work when that is done
6. turn the multi-compiled_params into a single parameters collection at DBAPI execution time and use cursor.execute().

**Comment by zzzeek** [comment]
updates, Oracle can do executemany + returning, but not multi VALUES.  So this continues that it looks like `connection.execute(insert().returning(), values)` should be enhanced to do "whatever it takes" to batch the values in.  So this starts looking again like the older plan, with a third flow to complement "do_execute()", "do_executemany()" that is something like "do_execute_insert_batch()", however, I think we still should limit a given values to a single "execute" or "executemany" on the DBAPI side at least for now.

so I continue to have to move away from my simplified idea at https://github.com/sqlalchemy/sqlalchemy/issues/6047#issuecomment-1081968916 as assuming multi-"VALUES" is too narrow an assumption.  at the same time, we should continue to get insert().values() to do execute time batching / caching as it will probably be inherent to how the feature actually works.

**Comment by zzzeek** [comment]
continuing, I think if one does `conn/session.execute(insert.retunring(), multivalues)`, that should automatically do the VALUES thing, and if not available, we raise.  This means it will fix #7907 as well.

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **main** branch:

**implement batched INSERT..VALUES () () for executemany** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4001
